# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


In [7]:
print('Гіпотези')
print('H0 = 123.7')
print('H1 > 123.7')
print('Тип правосторонній t-тест')

Гіпотези
H0 = 123.7
H1 > 123.7
Тип правосторонній t-тест


In [8]:
import numpy as np

# Вибірка
data = np.array([128, 135, 121, 142, 126, 151, 123])

# Розрахунок статистик
sample_mean = np.mean(data)
sample_std = np.std(data, ddof=1)  # ddof=1 для вибіркового стандартного відхилення
sample_size = len(data)

print("Середнє вибірки:", sample_mean)
print("Стандартне відхилення вибірки:", sample_std)
print("Розмір вибірки:", sample_size)

Середнє вибірки: 132.28571428571428
Стандартне відхилення вибірки: 10.980502200328935
Розмір вибірки: 7


In [9]:
from scipy import stats

mu0 = 123.7
alpha = 0.10

# t-статистика
t_stat = (sample_mean - mu0) / (sample_std / np.sqrt(sample_size))

# Критичне t-значення
t_critical = stats.t.ppf(1 - alpha, df=sample_size - 1)

# p-value (правосторонній тест)
p_value = 1 - stats.t.cdf(t_stat, df=sample_size - 1)

print("t-статистика:", t_stat)
print("Критичне t-значення:", t_critical)
print("p-value:", p_value)

t-статистика: 2.0687273144186475
Критичне t-значення: 1.4397557472577693
p-value: 0.042017654253682646


In [10]:
if t_stat > t_critical:
    print("t > t_critical → відхиляємо H0")
else:
    print("t <= t_critical → не відхиляємо H0")

if p_value < alpha:
    print("p-value < α → відхиляємо H0")
else:
    print("p-value >= α → не відхиляємо H0")

t > t_critical → відхиляємо H0
p-value < α → відхиляємо H0


t > t_critical → відхиляємо H0
p-value < α → відхиляємо H0
це означає, що є достатньо статистичних доказів, щоб стверджувати, що редизайн сайту збільшив середній час до покупки.

In [12]:
from scipy import stats

mu0 = 123.7
alpha = 0.05

# t-статистика
t_stat = (sample_mean - mu0) / (sample_std / np.sqrt(sample_size))

# Критичне t-значення
t_critical = stats.t.ppf(1 - alpha, df=sample_size - 1)

# p-value (правосторонній тест)
p_value = 1 - stats.t.cdf(t_stat, df=sample_size - 1)

print("t-статистика:", t_stat)
print("Критичне t-значення:", t_critical)
print("p-value:", p_value)

if t_stat > t_critical:
    print("t > t_critical → відхиляємо H0")
else:
    print("t <= t_critical → не відхиляємо H0")

if p_value < alpha:
    print("p-value < α → відхиляємо H0")
else:
    print("p-value >= α → не відхиляємо H0")

t-статистика: 2.0687273144186475
Критичне t-значення: 1.9431802805153018
p-value: 0.042017654253682646
t > t_critical → відхиляємо H0
p-value < α → відхиляємо H0


При зниженні рівня значущості до α = 0.05 є достатньо доказів, щоб стверджувати, що середній час до покупки після редизайну збільшився, тобто рішення не змінилося порівняно з α = 0.10 — редизайн справді подовжив час до покупки.

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [13]:
import statsmodels.api as sm

# Дані
n = 1200
x = 311
p0 = 0.23
alpha = 0.03

# Вибіркова пропорція
p_hat = x / n
print("Вибіркова пропорція:", p_hat)

Вибіркова пропорція: 0.25916666666666666


In [15]:
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

# Дані
n = 1200
x = 311
p0 = 0.23
alpha = 0.03

# Z-тест
z_stat, p_value = proportions_ztest(count=x, nobs=n, value=p0, alternative='larger')
print("Z-статистика:", z_stat)
print("p-value:", p_value)

# Висновок
if p_value < alpha:
    print("p-value < α → відхиляємо H0, пропорція збільшилась")
else:
    print("p-value >= α → не відхиляємо H0, недостатньо доказів")

# Довірчий інтервал
ci_low, ci_upp = proportion_confint(count=x, nobs=n, alpha=alpha, method='normal')
print(f"{(1-alpha)*100}% довірчий інтервал для пропорції: ({ci_low:.3f}, {ci_upp:.3f})")

Z-статистика: 2.3058316305222895
p-value: 0.010560021039195299
p-value < α → відхиляємо H0, пропорція збільшилась
97.0% довірчий інтервал для пропорції: (0.232, 0.287)


Це означає, що з імовірністю 97% справжня частка лежить у межах від 23.2% до 28.7%. Нижня межа майже дорівнює 23%, що в двосторонньому тесті не дозволило нам відхилити Н0, але при односторонньому ми оцінюємо це як значущість
